modules/benchmark/R/calc.metrics.R

```diff
@@ -1,83 +1,53 @@
-  vars.used.index <- which(format$vars$pecan_name %in% names(model)[!names(model)=="time"])
-  
-  # We know that the model output time is days since the beginning of the year.
-  # Make a column of years to supplement the column of days of the year.
-  years <- start_year:end_year
-  Diff <- diff(model$time)
-  n <- model$time[c(which(Diff < 0), length(model$time))]
-  y <- c()
-  for(i in 1:length(n)){
-    y <- c(y, rep(years[i], n[i]))
-  }
-  model$year <- y
-  model$posix <- strptime(paste(model$time, model$year), format = "%j %Y")
-  model_full = model
-  
-  # ---- INPUT DATA ---- #
-  
-  obvs <- load.data(data.path, format, start_year, end_year, site, vars.used.index, time.row)
-  dat_vars <- format$vars$pecan_name[vars.used.index]
-  obvs_full = obvs
-  
-  #########################
+calc.metrics <- function(model.bm, obvs.bm, var, metrics, start_year, end_year, bm, ens, model_run){
 
-  dat <- align.data(model_full, obvs_full, dat_vars, start_year, end_year)
+  dat <- align.data(model.bm, obvs.bm, var, start_year, end_year)
   
-  results <- as.data.frame(matrix(NA, nrow = length(metrics$name), ncol = length(dat_vars)+1))
-  colnames(results) <- c("metric", dat_vars)
+  results <- as.data.frame(matrix(NA, nrow = length(metrics$name), ncol = length(var)+1))
+  colnames(results) <- c("metric", var)
   rownames(results) <- metrics$name
   results$metric <- metrics$name
   
   file.sources = list.files("modules/benchmark/R/", pattern = "^metric.*", full.names = TRUE)
   for(i in 1:length(file.sources)){
     source(file.sources[i])
   }
-  
-  for(v in 1:length(dat_vars)){
-    metric_dat <- dat[,c(paste(dat_vars[v], c("m", "o"),sep = "." ),"posix")]
+
+    metric_dat <- dat[,c(paste(var, c("m", "o"),sep = "." ),"posix")]
     colnames(metric_dat)<- c("model","obvs","time")
     
     for(m in 1:length(metrics$name)){
       
       fcn <- paste0("metric.",metrics$name[m])
-      score <- as.character(do.call(fcn, args <- list(metric_dat,dat_vars[v])))
-      results[metrics$name[m],dat_vars[v]] <- score
-      
-      benchmarks_ensemble_id <- db.query(paste("SELECT id FROM benchmarks_ensembles where ensemble_id = ", ens$id),con )[[1]]
       
-      db.query(sprintf("INSERT into benchmarks_ensembles_scores (score, benchmarks_ensemble_id, benchmark_id, metric_id, created_at, updated_at) VALUES ('%s', %.0f, %.0f, %.0f, NOW(), NOW())",
-                       score, benchmarks_ensemble_id, bm$id, metrics$id[m]),con)
+      if(tail(unlist(strsplit(fcn, "[.]")),1) =="plot"){
+        filename = file.path(dirname(dirname(model_run)), 
+                             paste("benchmark",metrics$name[m],var,ens$id,"pdf", sep = "."))
+        do.call(fcn, args <- list(metric_dat,var,filename))
+        score <- filename
+        results[metrics$name[m],var] <- score
+      }else{
+        score <- as.character(do.call(fcn, args <- list(metric_dat,var)))
+        results[metrics$name[m],var] <- score
+      }
```

Comment:
Took me a while to infer the logic to what you're doing here (i.e. why plots were a special case, why the filename was calculated externally rather than returned by the function, etc). Seems like a good place to add some documentation.  I'd also suggest that you do capture what's returned by the metric.plot.* functions in case it throws back an error message.